0.23.0 blogpost (#278)
TeamNewPipe/website: _posts/2022-04-27-newpipe-0.23.0.md

- [@haklc](https://github.com/haklc) added a cool feature where you can change the audio pitch by semitones instead of percentage! This is a nifty tool for musically-inclined users, and also something fun to play with for everyone else. Don't worry, you can toggle a checkbox to revert to the percentage way of doing things.

- [@ktprograms](https://github.com/ktprograms) added a button to the queue menu in the main player which lets you add the current queue to a new or an existing playlist. Incidentally, this can also be used to 'convert' a remote playlist into a local NewPipe playlist, which you can then edit away to your heart's content. Isn't that such a nifty feature?

> - [@nv95](https://github.com/nv95) brings you another highly requested and anticipated feature: Notifications for new streams from subscribed channels! Now whenever your favourite YouTuber uploads a new video, NewPipe can send you a notification to let you know!

Reviewer: I'd move this to the beginning of the section, because it's the most important change and you wrote above:
> Where do we even start? Let's see... new stream notifications!

---
layout: post
title: "NewPipe 0.23.0 released: Behold the behemoth"
short: "NewPipe 0.23.0 released"
date: 2022-04-27 8:00:00 + 01:00
categories: [pinned, release]
author: <a href="https://github.com/opusforlife2">@Opusforlife</a> & <a href="https://github.com/poolitzer">@Poolitzer</a>
image: newpipe
excerpt_separator: <!-- more -->
---

This release got delayed by quite a bit, and so it just grew and grew and grew. You're getting a truly massive release to make up for the long wait. Be ready for a whopper of a blog post!

<!-- more -->

> Wow. Where do we even start? Let's see... new stream notifications! Now your phone can ping you even more frequently than before! Now you get seamless transition to and from background! Now there is no pause in the audio when switching between apps! YouTube/YouTube Music Mixes, the app suggests you these in the related items so that you can binge curated content to your hearts'... content. And finally, the popup player is usable again on Android 12!

Reviewer 2: I'd remove the "Now you get seamless transition to and from background!", since it is misleading, as people might think that *switching* players is instant, while it is not. "Now there is no pause in the audio when switching between apps!" is the correct thing to say. 

I would also remove the Mixes news, since YouTube has already broken them :-/

```suggestion
Wow. Where do we even start? Let's see... new stream notifications! Now your phone can ping you even more frequently than before! Now there is no pause in the audio when switching between apps! And finally, the popup player is usable again on Android 12!
```

Reviewer 3: > I would also remove the Mixes news, since YouTube has already broken them :-/

@opusforlife2 said it works for them, so it should be only an issue in the EU.

Reviewer 2: Oh, right, I will change my suggestions

This post also contains all the changes from the 0.22.0 Extractor version, which we were forced to release early as part of the hotfix to fix YouTube playback breaking yet again.

## New

- [@nv95](https://github.com/nv95) brings you a highly requested and anticipated feature: Notifications for new streams from subscribed channels! Now whenever your favourite YouTuber uploads a new video, NewPipe can send you a notification to let you know!

- [@TiA4f8R](https://github.com/TiA4f8R) brings you a second frequently requested and really awesome feature. Seamless switch to and from background! Now when you select "Minimize to background" in settings and minimise the app, NewPipe will pause the video stream and continue the audio stream without interruption! When you switch back to the app, you will see a black screen for a couple of seconds, after which the video will resume playing. (Note: there are certain (usually very old) videos for which this won't work. In those cases, the regular short pause in the audio will occur before playback resumes.)

> - [@Stypox](https://github.com/Stypox) made it so that YouTube Mixes and YouTube Music Mixes show up in related videos section. You're no longer forced to do the tedious job of having to share Mix URLs to NewPipe each time from another app to be able to play them! For those who have discovered upon reading this that NewPipe can play Mixes, congrats, and have fun! :)

Reviewer 2: ```suggestion
- [@Stypox](https://github.com/Stypox) ~~made it so that YouTube Mixes and YouTube Music Mixes show up in related videos section. You're no longer forced to do the tedious job of having to share Mix URLs to NewPipe each time from another app to be able to play them! For those who have discovered upon reading this that NewPipe can play Mixes, congrats, and have fun! :)~~ Well, actually YouTube already broke this feature by forcing users to accept cookies in order to show mixes in related items...
```

- [@haklc](https://github.com/haklc) added a cool feature where you can change the audio pitch by semitones instead of percentage! This is a nifty tool for musically-inclined users, and also something fun to play with for everyone else. Don't worry, you can toggle a checkbox to revert to the percentage way of doing things.

> - [@ktprograms](https://github.com/ktprograms) added a button to the queue menu in the main player which lets you add the current queue to a new or an existing playlist. Incidentally, this can also be used to 'convert' a remote playlist into a local NewPipe playlist, which you can then edit away to your heart's content. Isn't that such a nifty feature?

Reviewer 2: > Incidentally, this can also be used to 'convert' a remote playlist into a local NewPipe playlist, which you can then edit away to your heart's content. Isn't that such a nifty feature?

Since the "Add queue to playlist" button already existed for background and popup players (this PR just added it to the main player, too), you could already convert a remote playlist to a local one.

Author: The point doesn't say that only now you can convert a playlist, but you can with that suggestion.

Reviewer 2: Yeah, ok, I'm fine with it

- [@ktprograms](https://github.com/ktprograms) made the player remember the step size for the speed and pitch adjustment bars in the playback control dialog. Now the dialog doesn't default to the 25% step size each time you open it. Rather, it shows you the last used step size.

- [@Stypox](https://github.com/Stypox) ensured that all the relevant dialogs in the app uniformly show the "OK" and "Cancel" buttons, which are the default on Android, instead of "Yes" and "No" buttons.

- [@Stypox](https://github.com/Stypox) also improved the player UI and navigability for Android TV. This fixed several problems: unreachable buttons in the main player queue, incorrect background theming of the main player menus, and the player controls intermittently showing up below the queue menu. 

- [@SpinHit](https://github.com/SpinHit) added a confirmation dialog when deleting all files in the Download menu.

> - [@mauriciocolli](https://github.com/mauriciocolli) noticed that the app used duplicated icons for the night theme, which caused visibility problems for a lot of them. Since the original icons are vector drawables, the duplicated icons were deleted en masse, and instead, color attributes were used to match the original icons to the active theme.

Reviewer 2: Here I would also add a note about the fact that some small artifacts have probably been introduced, since this PR affected basically every place in the app where an icon was used. Some have already been reported, see https://github.com/TeamNewPipe/NewPipe/issues/8230#issuecomment-1108703650.

Author: ```suggestion
- [@mauriciocolli](https://github.com/mauriciocolli) noticed that the app used duplicated icons for the night theme, which caused visibility problems for a lot of them. Since the original icons are vector drawables, the duplicated icons were deleted en masse, and instead, color attributes were used to match the original icons to the active theme. We tried to catch every case where this breaks the icon, but it is likely we missed some. Feel free to report them in the repo (see the last header for more details on that).
```

- [@litetex](https://github.com/litetex) fixed the extraction of YouTube Shorts, which don't show the duration any longer after YouTube made a server-side change. Now the large number of "Not loaded" channels when refreshing the feed, as well as the annoyingly frequent error snackbar, won't bother you anymore.

- [@TiA4f8R](https://github.com/TiA4f8R) fixed the extraction of embeddable age-restricted videos on YouTube, as well as videos that provide a content warning and require user consent before playback. Keep in mind that YouTube can break the extraction of age-restricted videos again at any time, so enjoy it while it lasts.

- As a follow-up, [@TiA4f8R](https://github.com/TiA4f8R) also ensured that subtitles are extracted for age-restricted videos.

- The infamous 'n' parameter used by YouTube changed again, which caused throttling. That was fixed (again) by [@Stypox](https://github.com/Stypox).

- If you used external buttons (like on a Bluetooth headphone or speaker) to resume playback, the player controls UI would remain on screen forever, instead of fading out after a few moments. [@ktprograms](https://github.com/ktprograms) fixed this.

> - The update check would randomly fail while the app was in the background. [@TacoTheDank](https://github.com/TacoTheDank) updated some deprecated code to fix this.

Author: ```suggestion
- The app's update checking service would randomly fail while the app ran in the background, resulting in a crash. [@TacoTheDank](https://github.com/TacoTheDank) updated some deprecated code to fix this. Less crashes for you!
```

## Nerd Talk

- dependabot updates:
  - Bump actions/checkout from v2 to v3
  - Bump spotbugs-annotations from 4.5.3 to 4.6.0
  - Bump gson from 2.8.9 to 2.9.0
  - Bump actions/cache from 2 to 3
  - Bump rhino from 1.7.13 to 1.7.14

- [@litetex](https://github.com/litetex) refactored the `YoutubeCommentsExtractor` class to use the Java 8 Stream API, which greatly simplifies the code and reduces potential bugs.

- [@FireMasterK](https://github.com/FireMasterK) added support for extracting the channel subscriber count in the `StreamInfo` class.

- [@FireMasterK](https://github.com/FireMasterK) fixed a PeerTube test, and also removed the need for a new `CommentsInfo` instance each time one of the methods of this class had to be called, reducing the RAM burden caused by the unneeded instances.

- [@litetex](https://github.com/litetex) realised that our hardcoded SoundCloud ClientID was quickly after each release out of date, so there was no point in updating it again and again. It has been removed completely. Now we don't have to chase after a moving goalpost of an optimisation, which hardly benefited users. Thanks, litetex!

> - [@Stypox](https://github.com/Stypox) added the `MockOnly` JUnit 5 extension to all the YouTube extraction tests so that they work for Mixes. Moreover, the `DownloaderFactory` class's methods were made static, shaving off a little bit of CPU overhead each time that code is run.

Reviewer 2: ```suggestion
- [@Stypox](https://github.com/Stypox) added the `MockOnly` JUnit 5 extension so that a test can be instructed to only run when mocking the downloader. This was needed for YouTube Mixes. Moreover, the `DownloaderFactory` class's methods were made static, shaving off a little bit of CPU overhead each time that code is run.
```

- [@avently](https://github.com/avently) implemented a better equality check for queues, which ensures that going up the backstack via the Back button won't result in duplicates.

- [@TacoTheDank](https://github.com/TacoTheDank), in their LibraryMan avatar:
  - Gradle (and wrapper): 7.3.3 -> 7.4.2
  - Android Gradle Plugin: 7.0.4 -> 7.1.3
  - buildTools: 30.0.3 -> 31.0.0
  - Kotlin: 1.5.31 -> 1.6.20
  - ktlint: 0.43.2 -> 0.44.0
  - Checkstyle: 9.2.1 -> 9.3
  - Google Material: 1.4.0 -> 1.5.0 (which allowed one method to be updated to a much simpler version)
  - AndroidX:
    - LocalBroadcastManager: 1.0.0 -> 1.1.0
    - Media: 1.4.3 -> 1.5.0
    - Preference: 1.1.1 -> 1.2.0
    - ConstraintLayout: 2.1.2 -> 2.1.3
    - Room: 2.4.1 -> 2.4.2
  - GitHub Actions dependencies:
    - github-script: v5 -> v6
    - and all the rest: v2 -> v3

- While investigating some player UI layout issues, [@Stypox](https://github.com/Stypox) realised that the `large-land` XML layout file wasn't actually being used at all, and removed it, getting rid of some 800 lines of duplicated code!

> - [@TacoTheDank](https://github.com/TacoTheDank) migrated the app update checker from using the deprecated `IntentService` service to using a worker, by adding the AndroidX `Work` library. There was also some Kotlin-isation and refactoring involved.

Reviewer 4: Probably just a nitpick of mine lol, but I would change this up a little:

``
[@TacoTheDank](https://github.com/TacoTheDank) migrated the app update checker from using the deprecated `IntentService` service to using a worker (courtesy of the AndroidX Work library). This fixes the issue mentioned earlier where the update check would randomly fail while the app ran in the background, causing the app to crash. There was also some Kotlin-isation and refactoring involved.
``

Author: ```suggestion
- [@TacoTheDank](https://github.com/TacoTheDank) migrated the app update checker from using the deprecated `IntentService` service to using a worker, by adding the AndroidX `Work` library. There was also some Kotlin-isation and refactoring involved. Who doesn't like themselves some Kotlin.
```
